### PR TITLE
Client: Fix nonblocking-io zero by read

### PR DESCRIPTION
**Pull request by ffilz**
Since a return of 0 indicates a blocking read, if the length is 0, we must make an async completion call back and then return 0 indicating an async return rather than 0 bytes.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by petrutlucian94** [comment]
jenkins test windows

**Review by vshankar — COMMENTED** [review]
LGTM.  @ffilz You found this by a trivial 0byte read from nfs-gnaesha using the nonblocking-io interface?

**Comment by petrutlucian94** [comment]
For some reason, this breaks cephfs Windows support. The tests are failing consistently, also reproduced it locally:

```
ceph_test_dokan.DokanTests.test_mount_read_only
ceph_test_dokan.DokanTests.test_io
ceph_test_dokan.DokanTests.test_subfolders
ceph_test_dokan.DokanTests.test_find_files
ceph_test_dokan.DokanTests.test_move_file
ceph_test_dokan.DokanTests.test_max_path
ceph_test_dokan.DokanTests.test_set_eof
ceph_test_dokan.DokanTests.test_set_alloc_size
ceph_test_dokan.DokanTests.test_file_type
ceph_test_dokan.DokanTests.test_volume_info
ceph_test_dokan.DokanTests.test_get_free_space
ceph_test_dokan.DokanTests.test_file_timestamp
ceph_test_dokan.DokanTests.test_delete_disposition
ceph_test_dokan.DokanTests.test_create_dispositions
ceph_test_libcephfs_access
ceph_test_libcephfs
```

Let's defer merging this until we figure out the cause. I'll do some more digging.

**Comment by petrutlucian94** [comment]
``LibCephFS.BadArgument`` crashes on both platforms:

```
Thread 1 "ceph_test_libce" received signal SIGSEGV, Segmentation fault.
Client::_read_async (this=0x5555559bc600, f=0x5555557de100, off=5, len=<optimized out>, bl=0x7fffffffdea0, onfinish=0x0) at /home/ubuntu/workspace/ceph.linux/src/client/Client.cc:10950
10950       crf->complete(0);
(gdb) bt
#0  Client::_read_async (this=0x5555559bc600, f=0x5555557de100, off=5, len=<optimized out>,
    bl=0x7fffffffdea0, onfinish=0x0) at /home/ubuntu/workspace/ceph.linux/src/client/Client.cc:10950
#1  0x00007ffff7e6c0d5 in Client::_read (this=0x5555559bc600, f=0x5555557de100, offset=5, size=0,
    bl=0x7fffffffdea0, onfinish=0x0) at /home/ubuntu/workspace/ceph.linux/src/client/Client.cc:10773
#2  0x00007ffff7e6c749 in Client::read (this=0x5555559bc600, fd=10, buf=0x7fffffffdf80 " =zUUU",
    size=0, offset=5) at /home/ubuntu/workspace/ceph.linux/src/client/Client.cc:10524
#3  0x00005555555fc96f in LibCephFS_BadArgument_Test::TestBody (this=<optimized out>)
    at /home/ubuntu/workspace/ceph.linux/src/test/libcephfs/test.cc:1138
#4  0x00005555556fefef in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (location=0x55555571283a "the test body", method=<optimized out>, object=0x55555590e9e0)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2605
#5  testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (
    object=object@entry=0x55555590e9e0, method=<optimized out>,
    location=location@entry=0x55555571283a "the test body")
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2641
#6  0x00005555556f36f6 in testing::Test::Run (this=0x55555590e9e0)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2680
#7  testing::Test::Run (this=0x55555590e9e0)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2670
#8  0x00005555556f3855 in testing::TestInfo::Run (this=0x555555804d80)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2858
#9  testing::TestInfo::Run (this=0x555555804d80)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2831
#10 0x00005555556f3d19 in testing::TestSuite::Run (this=0x5555557fe020)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:3012
#11 testing::TestSuite::Run (this=0x5555557fe020)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2991
#12 0x00005555556f437f in testing::internal::UnitTestImpl::RunAllTests (this=0x5555557a3d20)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:5723
#13 0x00005555556ff5b7 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (
    location=0x555555715268 "auxiliary test code (environments or event listeners)",
    method=<optimized out>, object=0x5555557a3d20)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2605
#14 testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (
    object=0x5555557a3d20, method=<optimized out>,
    location=location@entry=0x555555715268 "auxiliary test code (environments or event listeners)")
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/src/gtest.cc:2641
#15 0x00005555556f3921 in testing::UnitTest::Run (
    this=0x5555557540a0 <testing::UnitTest::GetInstance()::instance>)
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/include/gtest/gtest.h:1429
#16 0x00005555555ebd57 in RUN_ALL_TESTS ()
    at /home/ubuntu/workspace/ceph.linux/src/googletest/googletest/include/gtest/gtest.h:2486
#17 main (argc=<optimized out>, argv=0x7fffffffe408)
    at /home/ubuntu/workspace/ceph.linux/src/test/libcephfs/main.cc:49

```

File: src/client/Client.cc [hunk 1]
```diff
@@ -10915,14 +10917,35 @@ int Client::_read_async(Fh *f, uint64_t off, uint64_t len, bufferlist *bl,
...
+    io_finish.reset(new C_Read_Async_Finisher(this, onfinish, f, in,
+                                              f->pos, off, len));
+  }
+
   // trim read based on file size?
-  if (off >= in->size)
-    return 0;
-  if (len == 0)
+  if ((off >= in->size) || (len == 0)) {
+    if (onfinish != nullptr)
```
Comment (petrutlucian94) [inline src/client/Client.cc:10932]: Did you actually mean ``if (onfinish == nulptr) { return 0; }``? At the moment, ``crf->complete(0)`` accesses a null pointer and crashes.

**Comment by ffilz** [comment]
> LGTM. @ffilz You found this by a trivial 0byte read from nfs-gnaesha using the nonblocking-io interface?

It was found by a kernel build. I tried to test with a trivial 0 byte read, but that may not be testing it properly.

I've lost track of where the issue was originally reported. The symptom is a hung client because the 0 byte read is being returned with an RC of 0 indicating "0 bytes" but 0 is taken to mean "expect async callback". So this patch is to convert a 0 byte immediate return into an async call back followed by RC 0 indicating "expect async callback" (which of course has already happened).

**Comment by ffilz** [comment]
So I need to find a more reliable 0 byte read that hangs without this patch...

Any suggestions? I tried creating a 0 byte file and then cat the file.

**Comment by ffilz** [comment]
OK, fixed the blunder. I'm still looking for a good way to validate.

**Comment by vshankar** [comment]
We probably need tests to validate the async paths. I'll create a tracker for this and have it added to our fs test suite.

**Review by vshankar — APPROVED** [review]

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#20-Sep-2023

**Comment by vshankar** [comment]
@ffilz please include this change with the backport you plan to do for nonblocking-io.

**Comment by ffilz** [comment]
Backport done.

**Comment by ffilz** [comment]
BTW, the original report for this was via nfs-ganesha github issue:

https://github.com/nfs-ganesha/nfs-ganesha/issues/988

**Comment by vshankar** [comment]
> Backport done.

Sorry, but where is the reef backport?

**Comment by vshankar** [comment]
> > Backport done.
> 
> Sorry, but where is the reef backport?

ping @ffilz

**Comment by ffilz** [comment]
Sorry, I did a downstream back port. I'm not clear where to do a Reef backport. It should be a trivial merge though.

Ceph backports are all new to me...

**Comment by vshankar** [comment]
> Sorry, I did a downstream back port. I'm not clear where to do a Reef backport. It should be a trivial merge though.
> 
> Ceph backports are all new to me...

Ah, ok. I'll go the reef backport.

**Pull request merged**